mlflow/tracking/context.py

```diff
@@ -0,0 +1,187 @@
+import os
+import sys
+import logging
+import itertools
+import warnings
+from abc import ABCMeta, abstractmethod
+
+import entrypoints
+from six.moves import reduce
+
+from mlflow.entities import SourceType
+from mlflow.utils import databricks_utils
+from mlflow.utils.mlflow_tags import (
+    MLFLOW_SOURCE_TYPE,
+    MLFLOW_SOURCE_NAME,
+    MLFLOW_GIT_COMMIT,
+    MLFLOW_DATABRICKS_WEBAPP_URL,
+    MLFLOW_DATABRICKS_NOTEBOOK_PATH,
+    MLFLOW_DATABRICKS_NOTEBOOK_ID
+)
+
+
+_logger = logging.getLogger(__name__)
+
+
+def _get_main_file():
+    if len(sys.argv) > 0:
+        return sys.argv[0]
+    return None
+
+
+def _get_source_name():
+    main_file = _get_main_file()
+    if main_file is not None:
+        return main_file
+    return "<console>"
+
+
+def _get_git_commit(path):
+    try:
+        import git
+    except ImportError as e:
+        _logger.warning(
+            "Failed to import Git (the Git executable is probably not on your PATH),"
+            " so Git SHA is not available. Error: %s", e)
+        return None
+    try:
+        if os.path.isfile(path):
+            path = os.path.dirname(path)
+        repo = git.Repo(path, search_parent_directories=True)
+        commit = repo.head.commit.hexsha
+        return commit
+    except (git.InvalidGitRepositoryError, git.GitCommandNotFound, ValueError, git.NoSuchPathError):
+        return None
+
+
+def _get_source_version():
+    main_file = _get_main_file()
+    if main_file is not None:
+        return _get_git_commit(main_file)
+    return None
+
+
+def _get_source_type():
+    return SourceType.LOCAL
+
+
+class ContextProvider(object):
+
+    __metaclass__ = ABCMeta
+
+    @abstractmethod
+    def in_context(self):
+        """
+        Determine if MLflow is running in this context.
+
+        :return: bool indicating if in this context
+        """
+        pass
+
+    @abstractmethod
+    def tags(self):
+        """
+        Generate context-specific tags.
+
+        :return: dict of tags
+        """
+        pass
+
+
+class DefaultContext(ContextProvider):
+    def in_context(self):
+        return True
+
+    def tags(self):
+        return {
+            MLFLOW_SOURCE_NAME: _get_source_name(),
+            MLFLOW_SOURCE_TYPE: _get_source_type()
+        }
+
+
+class GitContext(ContextProvider):
+    def in_context(self):
+        return _get_source_version() is not None
```

Comment:
nit: `_get_source_version` does disk I/O, so it seems like it would be ideal to only run it once. Consider caching the result, or unconditionally returning true (potentially putting it into `DefaultContext`) and only adding it to the returned tag list if it's non-None.

Reply:
I like the caching solution - I'll put something in.